### ISSUE TITLE
Difference between and_then and map

### ISSUE THREAD

**Issue by Timmmm** [seanmonstar/warp#594]
I'm having a little difficulty understanding the difference between the `map` and `and_then` filters.

For `Result` the difference seems to be that `and_then` replaces the entire `Result` with the function return value (if the original value is Ok) whereas `map` only replaces the Ok value. In other words you `map(foo)` is the same as `and_then(Ok(foo))`.

But in Warp they seem to be entirely different things because `and_then` is async and `map` isn't. Can you explain this? And is there a simple sync version of `and_then` that I can use instead of writing an async block without any async code?

**Comment by jxs** [comment]
Hi, yeah the difference in practice is their type signature, with `and_then` you can reject, also you don't need an async block, you can use [`futures::future::ok`](https://docs.rs/futures/0.3.5/futures/future/fn.ok.html) and [`futures::future::err`](https://docs.rs/futures/0.3.5/futures/future/fn.err.html)

**Comment by Timmmm** [comment]
Thanks for the reply! I still don't quite understand `and_then` is designed to take a future and `map` isn't? I mean, why don't I need to use `futures::future::ok()` in `map()`?

**Comment by Timmmm** [comment]
Sure, but can you explain why it was *designed* like that? It seems like it is mixing separate concerns (`Result<T>/T` return type, and sync/async). Shouldn't they both return futures? Or possibly have async and non-async versions like this:

* `and_then` - returns `Result<T>`
* `and_then_async` - returns `Future<Result<T>>`
* `map` - returns `T`
* `map_async` - returns `Future<T>`

**Comment by jxs** [comment]
they were designed to mimic [`futures::future::FutureExt::map`](https://docs.rs/futures/0.3.5/futures/future/trait.FutureExt.html#method.map) and [`futures::future::TryFutureExt`](https://docs.rs/futures/0.3.5/futures/future/trait.TryFutureExt.html#method.and_then)

**Comment by Timmmm** [comment]
Aha I see, it's `futures`'s fault!

So I thought it should be like this:

| Closure returns     | `Result<T>` | `Future<Result<T>>` |
|-|-|-|
| `T`                              |  `map` | `map` |
| `Result<T>`               | `and_then` | `and_then` |
| `Future<T>`               |  -        | `map_async` |
| `Future<Result<T>>` |  -       | `and_then_async` |

But it's actually like this (maybe, this stuff is really hard to follow):

| Closure returns     | `Result<T>` | `Future<Result<T>>` |
|-|-|-|
| `T`                              |  `map` | [`map_ok`][map_ok] |
| `Result<T>`               | `and_then` | Does not exist (a) |
| `Future<T>`               |  -        | Does not exist |
| `Future<Result<T>>` |  -       | [`and_then`][and_then] |

Even more confusing there is [another `map`][map] for `Future<Result<T>>` but unlike all the other functions here its function is called if the `Result<>` is `Ok` *or* `Err`, not just for `Ok`. I think you would implement (a) by something like `x.map(|x| x.map(...))`, where the outer `map` maps the entire `Result<T>` and then the inner one maps the `Ok` value.

Would you consider changing it to be consistent for V0.3?

[map]: https://docs.rs/futures/0.3.5/futures/future/trait.FutureExt.html#method.map
[map_ok]: https://docs.rs/futures/0.3.5/futures/future/trait.TryFutureExt.html#method.map_ok
[then]: https://docs.rs/futures/0.3.5/futures/future/trait.FutureExt.html#method.then
[and_then]: https://docs.rs/futures/0.3.5/futures/future/trait.TryFutureExt.html#method.and_then

**Comment by abreis** [comment]
I think I just hit a snag that's related to this. I needed to use `and_then` so I could call `.await` from within the closure, but its logic was infallible, so:
```rust
    .and_then(|data| async move {
        let response = data.get().await;
        Ok(warp::reply::json(&response))
    });
```
... failed, because `warp` couldn't infer the error type as `Rejection` (and rustc threw a semi-inscrutable error). This would work:
```rust
    if true { 
        Ok(warp::reply::json(&response)) 
    } else {
        Err(warp::reject::reject())
    }
```
But that's just not right. I was finally able to divine a better solution:
```rust
    Ok::<warp::reply::Json, warp::Rejection>(warp::reply::json(&client_list))
```

An infallible async method, as @Timmmm proposes, might be more ergonomic (though less discoverable?).

**Comment by drewkett** [comment]
Just so you know. You can use type inference to reduce some of the clutter on the return type. 

```
    Ok::<warp::reply::Json, warp::Rejection>(warp::reply::json(&client_list))
```

as

```
    Ok::<_, warp::Rejection>(warp::reply::json(&client_list))
```

I usually also have `use warp::Rejection;` at the top. So it becomes 

```
    Ok::<_, Rejection>(warp::reply::json(&client_list))
```

Not a huge difference, but I find the lack of clutter on the return types to be helpful.

**Comment by masklinn** [comment]
@abreis the alternative I found (before discovering @drewkett's version, I still have trouble with the turboshoals) was to move the callback to its own `async` function, which can be fully typed and solve the issue.
